semantic/constraint.go

```diff
@@ -0,0 +1,334 @@
+				left:  c.tenv[n.Argument],
+				right: Bool,
+			})
+		}
+	case *LogicalExpression:
+		// Inference Rule: Logical Expression
+		// ----------------------------------
+		// x = left (op) right
+		//
+		// -> typeof(x) = bool
+		// -> typeof(left) = bool
+		// -> typeof(right) = bool
+		//
+		// Logical expressions yield boolean types
+		// Logical operators act on boolean types
+		c.cons = append(c.cons,
+			Constraint{
+				left:  tv,
+				right: Bool,
+			},
+			Constraint{
+				left:  c.tenv[n.Left],
+				right: Bool,
+			},
+			Constraint{
+				left:  c.tenv[n.Right],
+				right: Bool,
+			},
+		)
+	case *BinaryExpression:
+		switch n.Operator {
+		// Inference Rule: Arithmetic Operators
+		// ------------------------------------
+		// x = a ( + , - , * , / ) b
+		//
+		// -> typeof(x) = typeof(a) = typeof(b)
+		//
+		// TODO: Only Int, UInt, Float, and String types supported
+		case ast.AdditionOperator:
+			c.cons = append(c.cons,
+				Constraint{
+					left:  tv,
+					right: c.tenv[n.Left],
+				},
+				Constraint{
+					left:  tv,
+					right: c.tenv[n.Right],
+				},
+				Constraint{
+					left:  c.tenv[n.Right],
+					right: c.tenv[n.Left],
+				},
+			)
+		case ast.SubtractionOperator:
+			c.cons = append(c.cons,
+				Constraint{
+					left:  tv,
+					right: c.tenv[n.Left],
+				},
+				Constraint{
+					left:  tv,
+					right: c.tenv[n.Right],
+				},
+				Constraint{
+					left:  c.tenv[n.Right],
+					right: c.tenv[n.Left],
+				},
+			)
+		case ast.MultiplicationOperator:
+			c.cons = append(c.cons,
+				Constraint{
+					left:  tv,
+					right: c.tenv[n.Left],
+				},
+				Constraint{
+					left:  tv,
+					right: c.tenv[n.Right],
+				},
+				Constraint{
+					left:  c.tenv[n.Right],
+					right: c.tenv[n.Left],
+				},
+			)
+		case ast.DivisionOperator:
+			c.cons = append(c.cons,
+				Constraint{
+					left:  tv,
+					right: c.tenv[n.Left],
+				},
+				Constraint{
+					left:  tv,
+					right: c.tenv[n.Right],
+				},
+				Constraint{
+					left:  c.tenv[n.Right],
+					right: c.tenv[n.Left],
+				},
+			)
+		// Inference Rules: Comparison Operators
+		// -------------------------------------
+		// x = left ( <, <=, >, >=, ==, != ) right
+		//
+		// -> typeof(x) = Bool
+		case ast.LessThanEqualOperator:
+			c.cons = append(c.cons,
+				Constraint{
+					left:  tv,
+					right: Bool,
+				})
+		case ast.LessThanOperator:
+			c.cons = append(c.cons,
+				Constraint{
+					left:  tv,
+					right: Bool,
+				})
+		case ast.GreaterThanEqualOperator:
+			c.cons = append(c.cons,
+				Constraint{
+					left:  tv,
+					right: Bool,
+				})
+		case ast.GreaterThanOperator:
+			c.cons = append(c.cons,
+				Constraint{
+					left:  tv,
+					right: Bool,
+				})
+		case ast.EqualOperator:
+			c.cons = append(c.cons,
+				Constraint{
+					left:  tv,
+					right: Bool,
+				})
+		case ast.NotEqualOperator:
+			c.cons = append(c.cons,
+				Constraint{
+					left:  tv,
+					right: Bool,
+				})
```

Comment:
Same here. Also should the operands must be the same type too?

Reply:
According to this, comparison operands may be of different types.

https://github.com/influxdata/flux/blob/4f007663549585e92c885e714da02dcb4a782c4a/semantic/binary_types.go#L35